skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,223 @@
+"""
+fisher_vector.py - Implementation of the Fisher vector encoding algorithm
+
+This module contains the source code for Fisher vector computation. The
+computation is separated into two distinct steps, which are called separately
+by the user, namely:
+
+learn_gmm: Used to estimate the GMM for all vectors/descriptors computed for
+           all examples in the dataset (e.g. estimated using all the SIFT
+           vectors computed for all images in the dataset, or at least a subset
+           of this).
+
+compute_fisher_vector: Used to compute the Fisher vector representation for a
+                       single set of descriptors/vector (e.g. the SIFT
+                       descriptors for a single image in your dataset, or
+                       perhaps a test image).
+
+Reference: Perronnin, F. and Dance, C. Fisher kernels on Visual Vocabularies
+           for Image Categorization, IEEE Conference on Computer Vision and
+           Pattern Recognition, 2007
+
+Origin Author: Dan Oneata (Author of the original implementation for the Fisher
+vector computation using scikit-learn and NumPy. Subsequently ported to
+scikit-image (here) by other authors.)
+"""
+import numpy as np
+
+
+class FisherVectorException(Exception):
+    pass
+
+
+class DescriptorException(FisherVectorException):
+    pass
+
+
+def learn_gmm(descriptors, num_modes, **kwargs):
```

Comment:
I presume this function is used only internall?  If so, rename to `_learn_gmm`.

Reply:
Nope. The standard way to implement is Fisher vector is to separate the GMM computation from the actual Fisher vector computation. The Fisher vector just depends on the derivatives of some generative probability model. Another reason to keep them separate is so you can compute the Fisher vector for new examples/images without needing to re-estimate the GMM.

vl-feat is probably the most popular implementation of the Fisher vector, and it also separates the GMM computation from the Fisher vector computation: https://www.vlfeat.org/overview/encodings.html

Reply:
OK, got it.